sycl/doc/extensions/SPIRV/SPV_INTEL_optimization_hints.asciidoc

```diff
@@ -0,0 +1,184 @@
+
+Notice
+------
+
+Copyright (c) 2020 Intel Corporation.  All rights reserved.
+
+Status
+------
+
+Working Draft
+
+This is a preview extension specification, intended to provide early access to a
+feature for review and community feedback. When the feature matures, this
+specification may be released as a formal extension.
+
+Because the interfaces defined by this specification are not final and are
+subject to change they are not intended to be used by shipping software
+products.
+
+Version
+-------
+
+[width="40%",cols="25,25"]
+|========================================
+| Last Modified Date | 2020-05-12
+| Revision           | G
+|========================================
+
+Dependencies
+------------
+
+This extension is written against the SPIR-V Specification,
+Version 1.5 Revision 2.
+
+This extension requires SPIR-V 1.0.
+
+Overview
+--------
+
+This extension adds *OpAssumeTrueINTEL* and *OpExpectINTEL* instructions that
+are mapped to ‘llvm.assume’ and ‘llvm.expect’ intrinsics respectively. That is
+required to save optimization hints through LLVM IR to SPIR-V transformation.
+
+Extension Name
+--------------
+
+To use this extension within a SPIR-V module, the following *OpExtension* must
+be present in the module:
+
+----
+OpExtension "SPV_INTEL_optimization_hints"
+----
+
+New Capabilities
+----------------
+This extension introduces a new capability:
+
+----
+OptimizationHintsINTEL
+----
+
+New Instructions
+----------------
+Instructions added under the *OptimizationHintsINTEL* capability:
+
+----
+OpExpectINTEL
+OpAssumeTrueINTEL
+----
+
+Token Number Assignments
+------------------------
+[width="45%",cols="30,15"]
+|===============================
+| OptimizationHintsINTEL | 5629
+| OpAssumeTrueINTEL      | 5630
+| OpExpectINTEL          | 5631
+|===============================
+
+Modifications to the SPIR-V Specification, Version 1.5
+------------------------------------------------------
+
+Capabilities
+~~~~~~~~~~~~
+
+Modify Section 3.31, "Capability", adding these rows to the Capability table:
+
+--
+[options="header"]
+|====
+2+^| Capability ^| Depends On
+| 5629 | *OptimizationHintsINTEL* +
+Allow to use OpAssumeTrueINTEL and OpExpectINTEL instructions |
+|====
+--
+
+Instructions
+~~~~~~~~~~~~
+
+In section 3.32.1, Miscellaneous Instructions, add a new instruction:
+
+[cols="3", width="100%"]
+|=====
+2+^|*OpAssumeTrueINTEL* +
+The instruction allows the optimizer to assume that the provided _Condition_ is
+always true whenever the control flow reaches the instruction. If the
+_Condition_ is violated during execution, the behavior is undefined.
+
+The type of _Condition_ must be _Boolean_.
+| Capability:
+*OptimizationHintsINTEL*
+
+| 2 | 5630 | Condition <id>
+|=====
+
+[cols="6", width="100%"]
+|=====
+5+^|*OpExpectINTEL* +
+
+Provides information for optimizers that the most probable value of _Value_ is
+_ExpectedValue_. This instruction can be safely replaced with an entity
+referenced by _Value_.
```

Comment:
I think we need to strengthen this description somewhat.  Is the intent that **OpExpectINTEL** behaves the same as **OpCopyObject** except that it also provides information about the most probable value of _Value_?

Reply:
> Is the intent that OpExpectINTEL behaves the same as OpCopyObject except that it also provides information about the most probable value of Value?

Yes, but OpExpectINTEL is more strict about result type.

Reply:
I still don't think this is sufficiently descriptive, and there is a typo now (`entityi`).

How about something like:

> This instruction behaves the same as *OpCopyObject* by making a copy of _Value_,
except it also provides information to the optimizer that the most probable
value of _Value_ is _ExpectedValue_.
>
> _Result Type_ must be a scalar or vector of _integer type_ or _Boolean type_.
> 
> The type of _Value_ and _ExpectedValue_ must be the same as _Result Type_.